File: ts/test/encode.test.ts
import { describe, test, expect } from "vitest";
import { encode, OverflowError, UnsupportedLengthError, MAX } from "../src/index.js";

describe("encode()", () => {
  describe("accepts number and bigint input", () => {
    test("encode(0, 6) and encode(0n, 6) produce identical output", () => {
      expect(encode(0, 6)).toBe(encode(0n, 6));
    });

    test("encode(1234567, 6) and encode(1234567n, 6) produce identical output", () => {
      expect(encode(1234567, 6)).toBe(encode(1234567n, 6));
    });
  });

  describe("defaults to length 6", () => {
    test("encode(0) defaults to length 6", () => {
      expect(encode(0)).toHaveLength(6);
    });

    test("encode(0) === '0A0000'", () => {
      expect(encode(0)).toBe("0A0000");
    });
  });

  describe("output properties", () => {
    test("output is always uppercase", () => {
      for (const n of [0, 1, 100, 999999, 1234567]) {
        const id = encode(n, 6);
        expect(id).toBe(id.toUpperCase());
      }
    });

    test("output length matches requested length", () => {
      expect(encode(0n, 6)).toHaveLength(6);
      expect(encode(0n, 7)).toHaveLength(7);
      expect(encode(0n, 8)).toHaveLength(8);
    });

    test("position 1 (0-indexed) is always an ALPHA character", () => {
      const ALPHA = "ABCDEFGHJKMNPQRSTVWXYZ";
      for (const n of [0n, 1n, 100n, 1234567n]) {
        const id = encode(n, 6);
        expect(ALPHA).toContain(id[1]);
      }
    });

    test("position 2 (0-indexed) is always a digit", () => {
      for (const n of [0n, 1n, 100n, 1234567n]) {
        const id = encode(n, 6);
        expect(id[2]).toMatch(/^\d$/);
      }
    });

    test("excluded characters I, L, O never appear in output", () => {
      const samples = [0n, 1n, 1000n, 1234567n, 100000000n, MAX[8] - 1n];
      for (const n of samples) {
        const id = encode(n, 8);
        expect(id).not.toMatch(/[ILO]/);
      }
    });
  });

  describe("boundary values", () => {
    test("encode(0n, 6) is valid", () => {
      expect(() => encode(0n, 6)).not.toThrow();
    });

    test("encode(MAX[6]-1, 6) is valid", () => {
      expect(() => encode(MAX[6] - 1n, 6)).not.toThrow();
    });

    test("encode(0n, 7) is valid", () => {
      expect(() => encode(0n, 7)).not.toThrow();
    });

    test("encode(MAX[7]-1, 7) is valid", () => {
      expect(() => encode(MAX[7] - 1n, 7)).not.toThrow();
    });

    test("encode(0n, 8) is valid", () => {
      expect(() => encode(0n, 8)).not.toThrow();
    });

    test("encode(MAX[8]-1, 8) is valid", () => {
      expect(() => encode(MAX[8] - 1n, 8)).not.toThrow();
    });
  });

  describe("overflow errors", () => {
    test("encode(MAX[6], 6) throws OverflowError", () => {
      expect(() => encode(MAX[6], 6)).toThrowError(OverflowError);
    });

    test("encode(MAX[7], 7) throws OverflowError", () => {
      expect(() => encode(MAX[7], 7)).toThrowError(OverflowError);
    });

    test("encode(MAX[8], 8) throws OverflowError", () => {
      expect(() => encode(MAX[8], 8)).toThrowError(OverflowError);
    });

    test("encode(-1n, 6) throws OverflowError", () => {
      expect(() => encode(-1n, 6)).toThrowError(OverflowError);
    });

    test("OverflowError is instance of RangeError", () => {
      try {
        encode(MAX[6], 6);
      } catch (e) {
        expect(e).toBeInstanceOf(RangeError);
      }
    });

    test("OverflowError message contains the violating value", () => {
      try {
        encode(MAX[6], 6);
      } catch (e) {
        expect((e as Error).message).toContain(String(MAX[6]));
      }
    });
  });

  describe("unsupported length errors", () => {
    const badLengths = [0, 1, 5, 9, 100];
    for (const l of badLengths) {
      test(`encode(0, ${l}) throws UnsupportedLengthError`, () => {
        // @ts-expect-error intentionally passing unsupported length
        expect(() => encode(0, l)).toThrowError(UnsupportedLengthError);
      });
    }
  });
});

File: dart/test/encode_test.dart
import 'package:odoid/odoid.dart';
import 'package:test/test.dart';

void main() {
  group('Encode', () {
    test('default length is 6', () {
      expect(OdoId.encode(0), equals('0A0000'));
      expect(OdoId.encode(0).length, equals(6));
    });

    for (final length in [6, 7, 8]) {
      test('output length matches requested ($length)', () {
        expect(OdoId.encode(0, length).length, equals(length));
      });

      test('encode(0, $length) is valid', () {
        expect(() => OdoId.encode(0, length), returnsNormally);
      });

      test('encode(MAX[$length]-1, $length) is valid', () {
        expect(() => OdoId.encode(kMax[length]! - 1, length), returnsNormally);
      });

      test('encode(MAX[$length], $length) throws OdoOverflowException', () {
        expect(() => OdoId.encode(kMax[length]!, length),
            throwsA(isA<OdoOverflowException>()));
      });
    }

    final samples = [0, 1, 100, 999999, 1234567];
    for (final n in samples) {
      test('output is uppercase for n=$n', () {
        final id = OdoId.encode(n, 6);
        expect(id, equals(id.toUpperCase()));
      });

      test('position 1 is always ALPHA char for n=$n', () {
        final ch = OdoId.encode(n, 6)[1];
        expect(kAlpha.contains(ch), isTrue,
            reason: 'pos 1 = "$ch" not in ALPHA');
      });

      test('position 2 is always a digit for n=$n', () {
        final ch = OdoId.encode(n, 6)[2];
        expect(RegExp(r'\d').hasMatch(ch), isTrue,
            reason: 'pos 2 = "$ch" not a digit');
      });
    }

    for (final n in [0, 1, 1000, 1234567, 100000000]) {
      test('excluded chars never appear for n=$n (length 8)', () {
        final id = OdoId.encode(n, 8);
        expect(id.contains('I'), isFalse, reason: '$id contains I');
        expect(id.contains('L'), isFalse, reason: '$id contains L');
        expect(id.contains('O'), isFalse, reason: '$id contains O');
      });
    }

    for (final length in [0, 1, 5, 9, 100]) {
      test('unsupported length $length throws UnsupportedLengthException', () {
        expect(() => OdoId.encode(0, length),
            throwsA(isA<UnsupportedLengthException>()));
      });
    }

    test('OdoOverflowException message contains violating value', () {
      try {
        OdoId.encode(kMax[6]!, 6);
        fail('expected OdoOverflowException');
      } on OdoOverflowException catch (e) {
        expect(e.toString(), contains(kMax[6]!.toString()));
      }
    });
  });
}
